IBM/operand-deployment-lifecycle-manager
# Operatorconfig controller

Controller to apply configuration from OperatorConfig onto CSVs
- triggered by OperandRequest CR or OperatorConfig CR
- searches through requested operators and checks registry for reference to any OperatorConfig
- applies configuration onto CSV's first deployment

This first version of OperatorConfig will be limited in scope to keep things simple - better to add things later than to remove or modify things.
- assumes configuration will only apply to CSV's first deployment. If there are multiple deployments, the other deployments do not get configured
- only three fields from CSV can be modified - replicas, affinity, and topologySpreadConstraints. This is different from initial design of just copying the entire CSV spec because there are certain fields in the CSV spec which are required to have a value, e.g. StrategyName. This poses problems for a user looking to configure just specific parts of the CSV because they would need to know all these required values, which are not necessarily relevant to their goal.

How to test:

Set DEV_REGISTRY and QUAY_REGISTRY to your own image registry before starting.

Use operator-sdk v1.29.0. Later versions have problems with `make bundle`

1. `make build-push-dev-image`
2. edit manager.yaml file to use built image
3. `make bundle-manifests`
4. `make build-push-bundle-image`
5. `make run-bundle`
6. Create default OperandRegistry and OperandConfig (install cs-operator to do this for you and then just remove cs-operator)
7. Edit default OperandRegistry's keycloak-operator to include `operatorConfig: test-operator-config` field
8. Create sample OperatorConfig provided
9. Create OperandRequest for keycloak-operator
10. Verify that Keycloak CSV is updated with what the OperatorConfig has, and verify that there are 2 Keycloak operator pods

Sample OperatorConfig

```yaml
apiVersion: operator.ibm.com/v1alpha1
kind: OperatorConfig
metadata:
  name: test-operator-config
spec:
  services:
    - name: keycloak-operator
      replicas: 2
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 90
            podAffinityTerm:
              topologyKey: topology.kubernetes.io/zone
              labelSelector:
                matchExpressions:
                - key: app.kubernetes.io/name
                  operator: In
                  values:
                  - cloud-native-postgresql
          - weight: 50
            podAffinityTerm:
              topologyKey: kubernetes.io/hostname
              labelSelector:
                matchExpressions:
                - key: app.kubernetes.io/name
                  operator: In
                  values:
                  - cloud-native-postgresql
      topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: topology.kubernetes.io/zone
          whenUnsatisfiable: ScheduleAnyway
          labelSelector:
            matchLabels:
              app.kubernetes.io/name: cloud-native-postgresql
        - maxSkew: 1
          topologyKey: topology.kubernetes.io/region
          whenUnsatisfiable: ScheduleAnyway
          labelSelector:
            matchLabels:
              app.kubernetes.io/name: cloud-native-postgresql
```
